### PR TITLE
[libiconv] Make built libraries relocatable

### PR DESCRIPTION
**Pull request by 375gnu**
This is required to build libintl (gettext) with iconv support (see #8484).

Also updated to the latest version 1.16.

### CODE REVIEW

File: ports/libiconv/CMakeLists.txt [hunk 1]
```diff
@@ -25,10 +25,29 @@ if(WIN32)
...
   string(REPLACE "#undef EILSEQ" "" _contents "${_contents}")
   string(REPLACE "#undef WORDS_LITTLEENDIAN" "#define WORDS_LITTLEENDIAN 1" _contents "${_contents}")
+  string(REPLACE "#undef ENABLE_RELOCATABLE" "#define ENABLE_RELOCATABLE 1" _contents "${_contents}")
   file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/config.h" "${_contents}")
 
   configure_file(libcharset/include/localcharset.h.build.in localcharset.h)
+  configure_file(libcharset/include/libcharset.h.in libcharset.h)
   configure_file(include/iconv.h.build.in iconv.h)
+
+  add_compile_options(/wd4018)
```
Comment (JackBoosY) [inline ports/libiconv/CMakeLists.txt]: Could you add judgment condition `if (MSVC)` to this line?

File: ports/libiconv/portfile.cmake [hunk 2]
```diff
@@ -7,20 +7,20 @@ endif()
 
 include(vcpkg_common_functions)
```
Comment (JackBoosY) [inline ports/libiconv/portfile.cmake]: We no longer need to include it, please remove this line.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by 375gnu** [comment]
@JackBoosY i fixed your remarks, will squash commits later whet the pr is approved

File: ports/libiconv/CMakeLists.txt [hunk 3]
```diff
@@ -25,10 +25,31 @@ if(WIN32)
...
+  configure_file(libcharset/include/libcharset.h.in libcharset.h)
   configure_file(include/iconv.h.build.in iconv.h)
+
+  if(MSVC)
+    add_compile_options(/wd4018)
+  endif()
+
+  add_definitions(
+    -D_CRT_SECURE_NO_WARNINGS
+    -DLIBDIR=\"c:/gettext\"
```
Comment (JackBoosY) [inline ports/libiconv/CMakeLists.txt]: We should pass these two paths from `vcpkg_configure_make` and set them here.

File: ports/libiconv/portfile.cmake [hunk 4]
```diff
@@ -5,22 +5,20 @@ if(VCPKG_CMAKE_SYSTEM_NAME AND NOT VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStor
     return()
```
Comment (JackBoosY) [inline ports/libiconv/portfile.cmake]: Can you replace `if(VCPKG_CMAKE_SYSTEM_NAME AND NOT VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")` to `if (NOT VCPKG_TARGET_IS_WINDOWS)`?

Thanks.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by 375gnu** [comment]
@JackBoosY done, please check

File: ports/libiconv/portfile.cmake [hunk 5]
```diff
@@ -29,6 +27,7 @@ file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
+    OPTIONS -DINSTALLDIR=\"c:/gettext\" -DLIBDIR=\"c:/gettext\"
```
Comment (JackBoosY) [inline ports/libiconv/portfile.cmake]: If this path means the final installation path of libiconv, it must be set to:
```
OPTIONS_DEBUG
-DINSTALLDIR=${CURRENT_PACKAGES_DIR}/debug/bin -DLIBDIR=${CURRENT_PACKAGES_DIR}/debug/lib
OPTIONS_RELEASE
-DINSTALLDIR=${CURRENT_PACKAGES_DIR}/bin -DLIBDIR=${CURRENT_PACKAGES_DIR}/lib
```
if this path means the installation path of getttext, it should be set to:
```
OPTIONS_DEBUG
-DINSTALLDIR=${CURRENT_INSTALLED_DIR}/debug/bin -DLIBDIR=${CURRENT_INSTALLED_DIR}/debug/lib
OPTIONS_RELEASE
-DINSTALLDIR=${CURRENT_INSTALLED_DIR}/bin -DLIBDIR=${CURRENT_INSTALLED_DIR}/lib
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/libiconv/portfile.cmake (hunk 5 above)
Comment (375gnu) [reply ports/libiconv/portfile.cmake]: Actually it doesn't matter what exact values is defined as the library made relocatable, i.e. able to load data not only from a fixed directory but from any other relative to the actual installation directory.

**Review by JackBoosY — APPROVED** [review]

**Review by ras0219-msft — APPROVED** [review]

**Pull request merged**